### ISSUE TITLE
Investigate app image support

### ISSUE THREAD

**Issue by pkgdemon** [furybsd/furybsd-livecd#164]
https://github.com/AppImage/AppImageKit/issues/98

@probonopd I see you are attached to the app image project.  I would be interested in shipping a solution like this with furybsd to possibly containerize all of its tooling separate from ports.  I am not oppressed to shipping a few more depends like Linux base, etc.  If you are using furybsd to investigate this I am interested in seeing how this can work for enhancing how we deliver and update furybsd assets cleanly on top of freebsd.

**Comment by probonopd** [comment]
Indeed.

We should distinguish between 2 different things:

1. Getting AppImages made for Linux to run on FreeBSD
2. Porting AppImage to FreeBSD or making FreeBSD-specific variation

**1. Getting AppImages made for Linux to run on FreeBSD**

For 1., at least the linux64 kernel module is needed on the FreeBSD side. Plus potentially a set of Linux libraries that are considered to come with all Desktop Linux distributions (loop-mounting an Ubuntu Live ISO might possibly do). 

Because one cannot really know what distributions ship, there is a tendency to bundle _everything_ in AppImages, down to and including glibc and ld-linux.so. As a result, such AppImages have a great chance to work on FreeBSD without needing any Linux libraries besides libfuse and glibc on the FreeBSD side. I have verified one such AppImage and it works on FreeBSD.

To run AppImages made for Linux on FreeBSD one currently has to unpack rather than mount them. Linux libfuse/squashfuse seems not to be able to use FreeBSD `/dev/fuse`, I was not yet able to find out whether this is a general design restriction or whether it is just a bug somewhere.

**2. Porting AppImage to FreeBSD or making FreeBSD-specific variation**

For 2., I think we should not just do a 1:1 port but deeply re-think "AppImage for FreeBSD" from the ground up, starting from the basic assumptions and constraints. AppImage was mainly created around real-live constraints of Linux and its various distributions. From what I have seen so far, not all of those apply to FreeBSD. For example, unlike with "Desktop Linux", in FreeBSD there is generally one way of doing things which is rather different from the Linux world where distributions seemingly cannot agree on anything. AppImage was made under the assumption that distributions cannot agree on anything and that they will not actively support AppImage. On the FuryBSD side, we might be able to create a much better, more deeply integrated experience if we can lift that restriction. So, an AppImage-like system for FuryBSD could do things slightly differently, and could potentially be more pleasant and with fewer downsides. I'm happy on brainstorming and working with you, and plan to experiment with a few things. I want to do things the FreeBSD way from the ground up rather than just blindly applying Linux-isms. For example, I'd like to think about how Jails, union-mounting, ZFS features like deduplication and possibly introducing some degree of interaction with the pkg system might be beneficial. It's too early for me to be concrete, I just need to play with FreeBSD more. And I'd be super happy to bounce ideas back and forth with people who have more FreeBSD experience than me.

So in summary, 1. is ready to be worked on immediately while 2. is still in the very, very early thinking stage.

Happy to hear your thoughts!

**Comment by Peter2121** [comment]
> **2. Porting AppImage to FreeBSD or making FreeBSD-specific variation**
> 
> For 2., I think we should not just do a 1:1 port but deeply re-think "AppImage for FreeBSD" from the ground up, starting from the basic assumptions and constraints. 

I would like to add this article to your sources of inspiration:
https://www.bsdstore.ru/en/xorg_in_jail.html

In an 'perfect world' I would like to download a jail image, import it, and start a GUI application :)

**Comment by probonopd** [comment]
> In an 'perfect world' I would like to download a jail image, import it, and start a GUI application :)

Yes, we are thinking along the same lines. Would it be even more ideal if want thent into the "jail image" was put together from existing `pkg`s? Maybe just throw some `pkg`s into a directory and have them `archivemount`ed and `unionfs`d and `chroot`ed automatically?

Really just thinking here, haven't played with that stuff yet. Just trying to make the point that "AppImage for BSD" might actually look quite different from AppImage for Linux. Better!

**Comment by Peter2121** [comment]
> Would it be even more ideal if want thent into the "jail image" was put together from existing `pkg`s? 
Sure, the main idea of using jails is to use everything that actually exists around. For example, today I can install any console application (for example, mc) inside a jail (using packages!), then I can mount my work directory inside the jail and use this application almost like it was installed on the main system. But if we need multiple application that should interact between them - it becomes more delicate ;)

**Comment by cmatomic** [comment]
here is a script to generate the functional .AppDir in FreeBSD I wanted to generate an appimage but the knowledge in FreeBSD is very limited, the result was a non-functional "appimage".

https://raw.githubusercontent.com/cmatomic/RecipesAppimage/master/FreeBSDappimage/chromiumAppimageV2.sh

**Comment by probonopd** [comment]
> 1. Getting AppImages made for Linux to run on FreeBSD

Thanks to @trasz there's a prototype kernel change that makes Linux squashfuse work, which is an important prerequisite: https://reviews.freebsd.org/D26959.
